### ISSUE TITLE
Avoid globally installing any dependencies

### ISSUE THREAD

**Issue by DanielSank** [labrad/scalabrad-web#195]
We'd like to be able to install packages like typescript locally.

npm users often globally install things like gulp and typescript because it puts binaries associated with those packages on the system PATH so that e.g. `$ gulp` works from the shell. This is convenient but it's not great that we have to make our entire system point at one particular version of gulp; what if different projects want different versions?

If you _locally_ install gulp in a particular npm package at `.../myproject` then in fact you get a symlink at `.../myproject/node_modules/.bin/gulp` which points to `.../myproject/node_modules/gulp/bin/gulp.js`. So you see we actually do have a binary we can use. There are two ways to get at this binary.
1. Put in on your PATH, e.g. via `bashrc`. We're already doing this to get at our `node` binary so this isn't the worst idea ever. However, we would like the repo to be self-sufficient, so if we were to do this I'd provide a `setup-my-shell.sh` script which does the appropriate PATH modifications.
2. Use `npm run`. This is a special program which looks for scripts defined in `package.json`'s `scripts` field. So for example, if `package.json` contains
   
   ```
   "scripts": {"gulp": ./node_modules/.bin/gulp}
   ```
   
   then `npm run gulp` does what we want. Even better though, `npm run` knows about the `.bin` directory, so you can just do this:
   
   ```
     "scripts": {"gulp": gulp}
   ```
   
   and it works the same. If we go this route, I think the only change we have to make (aside from the  documentation) is that we have to put e.g. `npm run typescript` in the gulp file.

Here's a somewhat more verbose version of what's written here:
https://firstdoit.com/no-need-for-globals-using-npm-dependencies-in-npm-scripts-3dfb478908#.wdwywwrfo

**Comment by maffoo** [comment]
I vote option 2, for sure.

I think we can avoid shelling out to typescript from inside the gulpfile by using the appropriate gulp plugin (gulp-typescript?) or maybe even shelling out to `npm run typescript` instead. The plugin is probably less hacky, but it means yet another dependency and as I recall when I tried it in the past the gulp-typescript plugin depended on a specific typescript version which was old, and in general you can't upgrade typescript until the plugin updates. Not a big deal now because the typescript version we need has been around for a while so the plugin probably works with it, but this is something to consider.

**Comment by DanielSank** [comment]
`package.json` says it wants typescript 1.8.0, which as far as I can tell is [what `gulp-typescript` uses by default](https://www.npmjs.com/package/gulp-typescript#typescript-version). Also, `gulp-typescript` is listed as a dev requirement in our `package.json` already, so I guess we're already using it...?

**Comment by DanielSank** [comment]
Note from installation: when I went to `$ npm install` the `client-js` directory (after local install of gulp etc.), I noticed the following

```
PhantomJS not found on PATH
Downloading https://github.com/Medium/phantomjs/releases/download/v1.9.19/phantomjs-1.9.8-linux-x86_64.tar.bz2
Saving to /tmp/phantomjs/phantomjs-1.9.8-linux-x86_64.tar.bz2
Receiving...
  [======================================--] 96%
Received 12854K total.
Extracting tar contents (via spawned process)
Removing /home/danielsank/src/scalabrad-web/client-js/node_modules/phantomjs/lib/phantom
Copying extracted folder /tmp/phantomjs/phantomjs-1.9.8-linux-x86_64.tar.bz2-extract-1466576819675/phantomjs-1.9.8-linux-x86_64 -> /home/danielsank/src/scalabrad-web/client-js/node_modules/phantomjs/lib/phantom
Writing location.js file
Done. Phantomjs binary available at /home/danielsank/src/scalabrad-web/client-js/node_modules/phantomjs/lib/phantom/bin/phantomjs
```

**Comment by DanielSank** [comment]
When I got to

`$ npm run bower install`

this happened:

```
Unable to find a suitable version for polymer, please choose one by typing one of the numbers below:
    1) polymer#^1.1.0 which resolved to 1.4.0 and is required by iron-component-page#1.1.7, iron-demo-helpers#1.2.3, iron-flex-layout#1.3.1, iron-flex-layout#1.3.1, iron-image#1.2.3, iron-list#1.3.4, iron-overlay-behavior#1.8.0, iron-pages#1.0.7, iron-scroll-threshold#1.0.1, iron-swipeable-container#1.0.2, marked-element#1.1.3, neon-animation#1.2.3, paper-badge#1.1.2, paper-button#1.0.12, paper-card#1.1.1, paper-checkbox#1.2.1, paper-dialog#1.0.4, paper-dialog-behavior#1.2.6, paper-drawer-panel#1.0.9, paper-dropdown-menu#1.2.2, paper-header-panel#1.1.6, paper-icon-button#1.1.2, paper-item#1.2.1, paper-listbox#1.1.2, paper-material#1.0.6, paper-menu#1.2.2, paper-menu-button#1.1.1, paper-progress#1.0.9, paper-radio-button#1.2.1, paper-radio-group#1.2.0, paper-ripple#1.0.6, paper-slider#1.0.11, paper-spinner#1.2.0, paper-tabs#1.6.2, paper-toggle-button#1.1.2, paper-tooltip#1.1.2
    2) polymer#^1.0.0 which resolved to 1.4.0 and is required by iron-a11y-announcer#1.0.4, iron-a11y-keys#1.0.5, iron-a11y-keys-behavior#1.1.3, iron-ajax#1.2.0, iron-autogrow-textarea#1.0.12, iron-checked-element-behavior#1.0.4, iron-dropdown#1.4.1, iron-fit-behavior#1.2.2, iron-form#1.0.16, iron-form-element-behavior#1.0.6, iron-icon#1.0.8, iron-icons#1.1.3, iron-iconset#1.0.4, iron-iconset-svg#1.0.9, iron-input#1.0.10, iron-jsonp-library#1.0.4, iron-label#1.0.2, iron-localstorage#1.0.5, iron-location#0.8.4, iron-media-query#1.0.8, iron-meta#1.1.1, iron-range-behavior#1.0.5, iron-resizable-behavior#1.0.3, iron-scroll-target-behavior#1.0.6, iron-signals#1.0.3, iron-test-helpers#1.2.5, iron-validatable-behavior#1.1.1, iron-validator-behavior#1.0.1, paper-dialog-scrollable#1.1.4, paper-fab#1.2.0, paper-scroll-header-panel#1.0.15, paper-styles#1.1.4, promise-polyfill#1.0.0
    3) polymer#^1.2.0 which resolved to 1.4.0 and is required by iron-behaviors#1.0.16, iron-selector#1.5.2, paper-input#1.1.11, paper-toolbar#1.1.5
    4) polymer#1.4.x which resolved to 1.4.0 and is required by labrad-browser
    5) polymer#>=1.0.8 which resolved to 1.4.0 and is required by polymer-ts#0.1.26
    6) polymer#^1.2.1 which resolved to 1.4.0 and is required by paper-behaviors#1.0.11
    7) polymer#^1.2.4 which resolved to 1.4.0 and is required by iron-menu-behavior#1.1.7
    8) polymer#^1.5.0 which resolved to 1.5.0 and is required by iron-collapse#1.1.0, iron-doc-viewer#1.0.15, paper-toast#1.3.0, prism-element#1.1.0

Prefix the choice with ! to persist it to bower.json
```
